src/pkcs11/pkcs11-global.c

```diff
@@ -306,12 +307,25 @@ CK_RV C_Initialize(CK_VOID_PTR pInitArgs)
 	in_finalize = 0;
 #endif
 
+        /* protect from nesting */
+	C_INITIALIZE_M_LOCK
+        nesting++;
+        if (nesting > 1) {
+            sc_log(context, "C_Initialize(): Nested init detected");
+            nesting--;
+	    C_INITIALIZE_M_UNLOCK
+            return CKR_GENERAL_ERROR;
+        }
+	C_INITIALIZE_M_UNLOCK
+        /* protect from nesting */
```

Comment:
the above 10 lines look like that's all what is needed to work around the original problem. Do we still need to use libctx then?

Reply:
Absolutely we do, see for example this comment that reiterates modules should use their own context:
https://github.com/openssl/openssl/issues/20315#issuecomment-1436554304

Quoting:
> Ideally, all modules would be using a custom context anyway

Reply:
thanks for the clearification